File: src/utils/transforms.py
"""Pandas-based aggregation helpers for the analytics dashboard."""
from __future__ import annotations

from typing import TYPE_CHECKING

import pandas as pd

if TYPE_CHECKING:
    pass


def to_orders_df(records: list[dict]) -> pd.DataFrame:
    df = pd.DataFrame(records)
    if df.empty:
        return df
    df["created_at"] = pd.to_datetime(df["created_at"])
    df["date"] = df["created_at"].dt.date
    return df


def revenue_over_time(df: pd.DataFrame, freq: str = "D") -> pd.DataFrame:
    """Resample orders to revenue time series. freq: 'D', 'W', 'ME'."""
    if df.empty:
        return pd.DataFrame(columns=["date", "revenue", "orders"])
    ts = (
        df.set_index("created_at")
        .resample(freq)
        .agg(revenue=("total_summ", "sum"), orders=("id", "count"))
        .reset_index()
        .rename(columns={"created_at": "date"})
    )
    return ts


def orders_by_status(df: pd.DataFrame) -> pd.DataFrame:
    if df.empty:
        return pd.DataFrame(columns=["status", "orders", "revenue"])
    return (
        df.groupby("status", dropna=False)
        .agg(orders=("id", "count"), revenue=("total_summ", "sum"))
        .reset_index()
        .sort_values("orders", ascending=False)
    )


def orders_by_manager(df: pd.DataFrame) -> pd.DataFrame:
    if df.empty:
        return pd.DataFrame(columns=["manager_name", "orders", "revenue", "avg_order"])
    grp = (
        df.groupby("manager_name", dropna=False)
        .agg(orders=("id", "count"), revenue=("total_summ", "sum"))
        .reset_index()
    )
    grp["avg_order"] = grp["revenue"] / grp["orders"].replace(0, pd.NA)
    return grp.sort_values("revenue", ascending=False)


def top_products(items_df: pd.DataFrame, n: int = 15) -> pd.DataFrame:
    if items_df.empty:
        return pd.DataFrame(columns=["product_name", "quantity", "revenue"])
    return (
        items_df.groupby("product_name", dropna=False)
        .agg(quantity=("quantity", "sum"), revenue=("revenue", "sum"))
        .reset_index()
        .sort_values("revenue", ascending=False)
        .head(n)
    )


def repeat_customers(df: pd.DataFrame) -> pd.DataFrame:
    if df.empty:
        return pd.DataFrame(columns=["customer_id", "customer_name", "orders", "revenue"])
    grp = (
        df.groupby(["customer_id", "customer_name"], dropna=False)
        .agg(orders=("id", "count"), revenue=("total_summ", "sum"))
        .reset_index()
    )
    return grp[grp["orders"] > 1].sort_values("orders", ascending=False)


# ---------------------------------------------------------------------------
# Funnel helpers
# ---------------------------------------------------------------------------

def _has_stage(popadal_series: pd.Series, code: str) -> pd.Series:
    """Boolean mask: True where the popadal list contains `code`."""
    return popadal_series.apply(lambda lst: code in lst if isinstance(lst, list) else False)


def funnel_stage_counts(
    df: pd.DataFrame,
    stages: list[tuple[str, str]],
) -> pd.DataFrame:
    """Count orders that passed through each stage (via popadal_v_statusy).

    stages: list of (status_code, popadal_code) tuples
    Returns DataFrame with columns: status_code, label, count, cr_pct
    """
    from src.utils.mappings import STATUS_LABELS
    if df.empty or "popadal_v_statusy" not in df.columns:
        return pd.DataFrame(columns=["status_code", "label", "count", "cr_pct"])

    base = len(df)
    rows = []
    for status_code, popadal_code in stages:
        cnt = int(_has_stage(df["popadal_v_statusy"], popadal_code).sum())
        rows.append({
            "status_code": status_code,
            "label": STATUS_LABELS.get(status_code, status_code),
            "count": cnt,
            "cr_pct": round(cnt / base * 100, 1) if base else 0.0,
        })
    return pd.DataFrame(rows)


def funnel_over_time(
    df: pd.DataFrame,
    stages: list[tuple[str, str]],
    freq: str = "D",
) -> pd.DataFrame:
    """Multi-line time series: for each period bucket, count orders per stage.

    Returns a wide DataFrame with one column per stage label + a 'date' column.
    """
    from src.utils.mappings import STATUS_LABELS
    if df.empty or "popadal_v_statusy" not in df.columns:
        return pd.DataFrame()

    ts = df.set_index("created_at").resample(freq)

    result = pd.DataFrame(index=ts.groups.keys())
    result.index.name = "date"

    for status_code, popadal_code in stages:
        label = STATUS_LABELS.get(status_code, status_code)
        mask = _has_stage(df["popadal_v_statusy"], popadal_code)
        col = df[mask].set_index("created_at").resample(freq).size().rename(label)
        result = result.join(col, how="left")

    result = result.fillna(0).astype(int).reset_index()
    return result


def financial_kpis(df: pd.DataFrame) -> dict:
    """Compute Total Sales, MRR, SARPU, Total Refunded, Net Total."""
    from src.utils.mappings import PAID_STATUS_CODE, REFUND_STATUS_CODE

    if df.empty or "popadal_v_statusy" not in df.columns:
        return {"total_sales": 0, "mrr": 0.0, "sarpu": 0.0, "refunded": 0.0, "net_total": 0.0}

    paid_mask   = _has_stage(df["popadal_v_statusy"], PAID_STATUS_CODE)
    refund_mask = _has_stage(df["popadal_v_statusy"], REFUND_STATUS_CODE)

    paid_df   = df[paid_mask]
    refund_df = df[refund_mask]

    total_sales = int(paid_mask.sum())
    sarpu       = float(paid_df["cf_first_payment"].sum())

    periods = paid_df["cf_payment_period"].replace(0, pd.NA)
    mrr     = float((paid_df["cf_first_payment"] / periods).sum())

    refunded  = float(refund_df["cf_refunded"].sum())
    net_total = sarpu - refunded

    return {
        "total_sales": total_sales,
        "mrr":         mrr,
        "sarpu":       sarpu,
        "refunded":    refunded,
        "net_total":   net_total,
    }


def platforms_breakdown(df: pd.DataFrame) -> pd.DataFrame:
    """Count orders by previous platform (cf_prev_platform)."""
    if df.empty or "cf_prev_platform" not in df.columns:
        return pd.DataFrame(columns=["platform", "count"])
    return (
        df.groupby("cf_prev_platform", dropna=False)
        .size()
        .reset_index(name="count")
        .rename(columns={"cf_prev_platform": "platform"})
        .sort_values("count", ascending=False)
    )

File: web/src/lib/transforms.ts
import dayjs from "dayjs";
import weekOfYear from "dayjs/plugin/weekOfYear";
import type { OrderRecord, ItemRecord } from "./flatten";
import {
  PAID_STATUS_CODE,
  REFUND_STATUS_CODE,
  PLATFORM_LABELS,
  SECTOR_LABELS,
  type StageEntry,
} from "./mappings";

dayjs.extend(weekOfYear);

export type Freq = "D" | "W" | "ME";

// ---------------------------------------------------------------------------
// Date bucketing
// ---------------------------------------------------------------------------

function bucket(iso: string, freq: Freq): string {
  const d = dayjs(iso);
  if (freq === "D") return d.format("YYYY-MM-DD");
  if (freq === "W") return d.startOf("week").format("YYYY-MM-DD");
  return d.startOf("month").format("YYYY-MM-DD");
}

// ---------------------------------------------------------------------------
// General analytics transforms
// ---------------------------------------------------------------------------

export interface TsPoint {
  date: string;
  revenue: number;
  orders: number;
}

export function revenueOverTime(records: OrderRecord[], freq: Freq): TsPoint[] {
  const map = new Map<string, TsPoint>();
  for (const r of records) {
    const d = bucket(r.createdAt, freq);
    const existing = map.get(d) ?? { date: d, revenue: 0, orders: 0 };
    map.set(d, { date: d, revenue: existing.revenue + r.totalSumm, orders: existing.orders + 1 });
  }
  return [...map.entries()].sort((a, b) => a[0].localeCompare(b[0])).map(([, v]) => v);
}

export interface StatusRow {
  status: string;
  orders: number;
  revenue: number;
}

export function ordersByStatus(records: OrderRecord[]): StatusRow[] {
  const map = new Map<string, StatusRow>();
  for (const r of records) {
    const s = r.status ?? "(sin estado)";
    const existing = map.get(s) ?? { status: s, orders: 0, revenue: 0 };
    map.set(s, { status: s, orders: existing.orders + 1, revenue: existing.revenue + r.totalSumm });
  }
  return [...map.values()].sort((a, b) => b.orders - a.orders);
}

export interface ManagerRow {
  managerId: number | null;
  managerName: string;
  orders: number;
  revenue: number;
  avgOrder: number;
}

export const NO_MANAGER_KEY = "__no_manager__";

export function ordersByManager(records: OrderRecord[]): ManagerRow[] {
  const map = new Map<string, ManagerRow>();
  for (const r of records) {
    const key = r.managerSd || r.managerName || NO_MANAGER_KEY;
    const existing = map.get(key) ?? { managerId: r.managerId, managerName: key, orders: 0, revenue: 0, avgOrder: 0 };
    const updated = { ...existing, orders: existing.orders + 1, revenue: existing.revenue + r.totalSumm };
    updated.avgOrder = updated.revenue / updated.orders;
    map.set(key, updated);
  }
  return [...map.values()].sort((a, b) => b.revenue - a.revenue);
}

export interface ProductRow {
  productName: string;
  quantity: number;
  revenue: number;
}

export function topProducts(items: ItemRecord[], n = 15): ProductRow[] {
  const map = new Map<string, ProductRow>();
  for (const item of items) {
    const key = item.productName || "(sin nombre)";
    const existing = map.get(key) ?? { productName: key, quantity: 0, revenue: 0 };
    map.set(key, { productName: key, quantity: existing.quantity + item.quantity, revenue: existing.revenue + item.revenue });
  }
  return [...map.values()].sort((a, b) => b.revenue - a.revenue).slice(0, n);
}

export interface RepeatCustomer {
  customerId: number | null;
  customerName: string;
  orders: number;
  revenue: number;
}

export function repeatCustomers(records: OrderRecord[]): RepeatCustomer[] {
  const map = new Map<string, RepeatCustomer>();
  for (const r of records) {
    const key = String(r.customerId ?? r.customerName ?? "?");
    const existing = map.get(key) ?? { customerId: r.customerId, customerName: r.customerName, orders: 0, revenue: 0 };
    map.set(key, { ...existing, orders: existing.orders + 1, revenue: existing.revenue + r.totalSumm });
  }
  return [...map.values()].filter((c) => c.orders > 1).sort((a, b) => b.orders - a.orders);
}

// ---------------------------------------------------------------------------
// Funnel transforms
// ---------------------------------------------------------------------------

function hasStage(record: OrderRecord, code: string): boolean {
  return record.popadalVStatusy.includes(code);
}

export interface StageRow {
  statusCode: string;
  label: string;
  count: number;
  crPct: number;
}

export function funnelStageCounts(records: OrderRecord[], stages: StageEntry[], base?: number): StageRow[] {
  const denominator = base ?? records.length;
  return stages.map(([label, popadalCode]) => {
    const count = records.filter((r) => hasStage(r, popadalCode)).length;
    return {
      statusCode: popadalCode,
      label,
      count,
      crPct: denominator > 0 ? Math.round((count / denominator) * 100) : 0,
    };
  });
}

export interface FunnelTsPoint {
  date: string;
  [stageName: string]: string | number;
}

export function funnelOverTime(
  records: OrderRecord[],
  stages: StageEntry[],
  freq: Freq
): FunnelTsPoint[] {
  const bucketMap = new Map<string, Map<string, number>>();

  for (const r of records) {
    const d = bucket(r.createdAt, freq);
    if (!bucketMap.has(d)) bucketMap.set(d, new Map());
    const stageMap = bucketMap.get(d)!;
    for (const [label, popadalCode] of stages) {
      if (hasStage(r, popadalCode)) {
        stageMap.set(label, (stageMap.get(label) ?? 0) + 1);
      }
    }
  }

  const stageLabels = stages.map(([label]) => label);
  return [...bucketMap.entries()]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([date, stageMap]) => {
      const point: FunnelTsPoint = { date };
      for (const label of stageLabels) {
        point[label] = stageMap.get(label) ?? 0;
      }
      return point;
    });
}

export interface FinancialKpis {
  totalSales: number;
  mrr: number;
  sarpu: number;
  refunded: number;
  netTotal: number;
}

export function financialKpis(records: OrderRecord[]): FinancialKpis {
  // Use popadal code when available; fall back to cfFirstPayment > 0 when the
  // stage wasn't recorded in popadal_v_statusy (e.g. orders pre-dating the field)
  const paid = records.filter((r) => hasStage(r, PAID_STATUS_CODE) || r.cfFirstPayment > 0);
  const refund = records.filter((r) => hasStage(r, REFUND_STATUS_CODE) || r.cfRefunded > 0);

  const sarpu = paid.reduce((s, r) => s + r.cfFirstPayment, 0);
  const mrr = paid.reduce((s, r) => s + (r.cfPaymentPeriod > 0 ? r.cfFirstPayment / r.cfPaymentPeriod : 0), 0);
  const refunded = refund.reduce((s, r) => s + r.cfRefunded, 0);

  return {
    totalSales: paid.length,
    mrr,
    sarpu,
    refunded,
    netTotal: sarpu - refunded,
  };
}

export interface PlatformRow {
  platform: string;
  count: number;
}

const NO_DATA = "No rellenado";

export function platformsBreakdown(records: OrderRecord[]): PlatformRow[] {
  const map = new Map<string, number>();
  for (const r of records) {
    const raw = (r.cfPrevPlatform ?? "").trim();
    const label = raw ? (PLATFORM_LABELS[raw.toLowerCase()] ?? raw) : NO_DATA;
    const key = label === "No rellenado" ? NO_DATA : label;
    map.set(key, (map.get(key) ?? 0) + 1);
  }
  return [...map.entries()]
    .map(([platform, count]) => ({ platform, count }))
    .sort((a, b) => a.platform === NO_DATA ? 1 : b.platform === NO_DATA ? -1 : b.count - a.count);
}

export function sectorBreakdown(records: OrderRecord[]): PlatformRow[] {
  const map = new Map<string, number>();
  for (const r of records) {
    const raw = (r.cfSector ?? "").trim();
    const name = raw ? (SECTOR_LABELS[raw.toLowerCase()] ?? raw) : NO_DATA;
    const key = name === "No rellenado" ? NO_DATA : name;
    map.set(key, (map.get(key) ?? 0) + 1);
  }
  return [...map.entries()]
    .map(([platform, count]) => ({ platform, count }))
    .sort((a, b) => a.platform === NO_DATA ? 1 : b.platform === NO_DATA ? -1 : b.count - a.count);
}
